### PR TITLE
cleanup: tast.fs --> TypedTree.fs and other items

### PR DESCRIPTION
**Pull request by dsyme**
This is cleanup

* [x] rename `AbstractSyntax.fs` to `SyntaxTree.fs` etc.
* [x] rename `tast.fs` to `TypedTree.fs`, `tastops.fs` to `TypedTreeOps.fs` etc.
* [x] adopt `module rec` at the top of `TypedTree.fs` to make the domain model presentation clearer
* [x] split out `TypedTreeBasics.fs` to contain the somewhat hodge-podge of functions that were in `tast.fs`.  These are the ones needed for `TcGlobals.fs` and we can gradually either move them into the domain model `TypedTree.fs` or into `TcGloblals.fs` or `TypedTreeOps.fs`
* [x] move the TAST construction functions into `Construct` alongside the others already there.
* [x] miscellaneous indentation and whitespace cleanup
* [x] document more things in `TypedTree.fs`
* [x] remove unused `FullAbstraction` exception

The key aim is that `TypedTree.fs` should be much clearer and visually possible to verify that everything is clean and properly documentd

### CODE REVIEW

**Comment by auduchinok** [comment]
Shouldn't `AbstractSyntax` and `TypedAST` be consistent here? I've always liked how easy it was to see a connection between `Ast` and `Tast` names.

**Comment by dsyme** [comment]
> Shouldn't AbstractSyntax and TypedAST be consistent here? I've always liked how easy it was to see a connection between Ast and Tast names.

Yeah I know, though I don't have a specific suggestion.  Any ideas?

The two trees are actually quite different
* AST is basically immutable - TAST uses mutability for linking, pickling and inference
* AST is really a tree, TAST is really a graph/DAG

However I know what you mean, it was a comforting thing

**Comment by auduchinok** [comment]
> Yeah I know, though I don't have a specific suggestion. Any ideas?

The only idea I've got so far is to rename `AbstractSyntax` back to `Ast`. Though, I doubt you'll like it much. :)

**Comment by dsyme** [comment]
> The only idea I've got so far is to rename AbstractSyntax back to Ast. Though, I doubt you'll like it much. :)

Option 1: Ast.fs, AstOps.fs, TypedAst.fs, TypedAstOps.fs

Option 2: AbstractSyntax.fs, AbstractSyntaxOps.fs, TypedAbstractSyntax.fs, TypedAbstractSyntaxOps.fs

Option 3: SyntaxTree.fs, SyntaxTreeOps.fs, TypedTree.fs, TypedTreeOps.fs


What's your preference order?

**Comment by auduchinok** [comment]
I definitely like 1 and 3 the most. It seems like I'm more used to something like 1, but at the same time 3 looks really good too.

**Comment by KevinRansom** [comment]
@dsyme this is an enormous set of changes.  Could you perhaps pick a focused set of cleanup to submit, and remember we just merged one 25 file cleanup session.  And space out the cleanups over time, so that we can handle the branch merges cleanly for each.  It also gives PR owners a chance to handle their own conflicts with master more reasonably too.

In general cleanup is appreciated, but we need to keep our eye on the fact that we have, shipping branches and the fcs repo to consider.

Thanks

Kevin

**Comment by dsyme** [comment]
Ok option 3 is good.

**Comment by dsyme** [comment]
> @dsyme this is an enormous set of changes. 

I think it's not as big as it looks? Just +2000/-2000.  Most the files are just a few lines of changes due to the rename (which necessarily affects about 100 files), plus reordering the `open` declarations

File: src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.ProjectFile.fs [hunk 1]
```diff
@@ -11,56 +11,63 @@ open System.Runtime.CompilerServices
...
-    RestoreSources:string
-    Script:string }
-
+type PackageReference =
+    {
+      Include:string
+      Version:string
+      RestoreSources:string
+      Script:string
+    }
```
Comment (auduchinok) [inline src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.ProjectFile.fs]: It seems like the following style is most commonly used accross the compiler code.
```suggestion
type PackageReference =
    { Include:string
      Version:string
      RestoreSources:string
      Script:string }
```

Comment (dsyme) [reply src/fsharp/FSharp.DependencyManager.Nuget/FSharp.DependencyManager.ProjectFile.fs]: This style has been most common but doesn't work well when `///` comments and newline are added - see Typoe

We should move systematically to the second style and add `///` comments everywhere.  No more undocumented code.

**Comment by dsyme** [comment]
@KevinRansom I'm done with the cleanup pass now, can we push this one through?  Am happy to help with any problematic integrations.  I think it won't be too bad - the files being changed are very stable.  The things most likely affected are my RFC PRs but I'll do those

**Pull request merged**